packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js

```diff
@@ -0,0 +1,573 @@
+const amendmentProductPurposeKeys = [
+  'STRUCTURE',
+  'MOISTURE_RETENTION',
+  'NUTRIENT_AVAILABILITY',
+  'PH',
+  'OTHER',
+];
+const amendmentMethodKeys = [
+  'BROADCAST',
+  'BANDED',
+  'FURROW_HOLE',
+  'SIDE_DRESS',
+  'FERTIGATION',
+  'FOLIAR',
+  'OTHER',
+];
+const furrowHoleDepthUnits = ['cm', 'in'];
+const metricWeightUnits = ['g', 'kg', 'mt'];
+const imperialWeightUnits = ['oz', 'lb', 't'];
+const weightUnits = [...metricWeightUnits, ...imperialWeightUnits];
+const applicationRateWeightUnits = [
+  'g/m2',
+  'lb/ft2',
+  'kg/m2',
+  't/ft2',
+  'mt/m2',
+  'oz/ft2',
+  'g/ha',
+  'lb/ac',
+  'kg/ha',
+  't/ac',
+  'mt/ha',
+  'oz/ac',
+];
+const metricVolumeUnits = ['ml', 'l'];
+const imperialVolumeUnits = ['fl-oz', 'gal'];
+const volumeUnits = [...metricVolumeUnits, ...imperialVolumeUnits];
+const applicationRateVolumeUnits = [
+  'l/m2',
+  'gal/ft2',
+  'ml/m2',
+  'fl-oz/ft2',
+  'l/ha',
+  'gal/ac',
+  'ml/ha',
+  'fl-oz/ac',
+];
+
+export const up = async function (knex) {
+  // Create fertliser types table
+  await knex.schema.createTable('soil_amendment_fertiliser_type', (table) => {
+    table.increments('id').primary();
+    table.string('key').notNullable();
+  });
+  // Add fertiliser types
+  for (const key of soilAmendmentFertiliserTypeKeys) {
+    await knex('soil_amendment_fertiliser_type').insert({
+      key,
+    });
+  }
+
+  await knex.schema.createTable('soil_amendment_product', (table) => {
+    table.integer('product_id').references('product_id').inTable('product').primary();
+    table
+      .integer('soil_amendment_fertiliser_type_id')
+      .references('id')
+      .inTable('soil_amendment_fertiliser_type')
+      .nullable();
+    table.decimal('n');
+    table.decimal('p');
+    table.decimal('k');
+    table.decimal('calcium');
+    table.decimal('magnesium');
+    table.decimal('sulfur');
+    table.decimal('copper');
+    table.decimal('manganese');
+    table.decimal('boron');
+    table.enu('elemental_unit', elementalUnits);
+    table.decimal('ammonium');
+    table.decimal('nitrate');
+    table.enu('molecular_compounds_unit', molecularCompoundsUnits);
+    table.check(
+      `(COALESCE(${elements.join(
+        ', ',
+      )}) IS NULL AND elemental_unit IS NULL) OR (COALESCE(${elements.join(
+        ', ',
+      )}) IS NOT NULL AND elemental_unit IS NOT NULL)`,
+      [],
+      'elemental_unit_check',
+    );
+    table.check(
+      `elemental_unit != 'percent' OR (elemental_unit = 'percent' AND (${elements
+        .map((element) => `COALESCE(${element}, 0)`)
+        .join(' + ')}) <= 100)`,
+      [],
+      'elemental_percent_check',
+    );
+    table.check(
+      `(COALESCE(${compounds.join(
+        ', ',
+      )}) IS NULL AND molecular_compounds_unit IS NULL) OR (COALESCE(${compounds.join(
+        ', ',
+      )}) IS NOT NULL AND molecular_compounds_unit IS NOT NULL)`,
+      [],
+      'molecular_compounds_unit_check',
+    );
+  });
```

Comment:
Which table has the `moisture/dry matter content`?
Here is the check required that I mentioned earlier (excerpt from the [user story](https://lite-farm.atlassian.net/browse/LF-4213)):
```
Moisture content and Dry matter content (Optional) 
Interaction: Both fields must add up to 100%. When one field is filled, the other field automatically updates to ensure the total is 100%.
```

Reply:
@Duncan-Brain it should live on `soil_amendment_product`, right?

Reply:
Solid oversight on my part. Yes it looks like `soil_amendment_product` in the figma screens. I agree with it I think.

Reply:
Should we check if `moisture_content_percent` <= 100?

Reply:
Right, for some reason I thought since we don't need to store dry -- but it is still useful yah.